pageserver/src/tenant/timeline/walreceiver/connection_manager.rs

```diff
@@ -409,12 +414,33 @@ impl ConnectionManagerState {
                     ctx,
                     node_id,
                 )
-                .await
-                .context("walreceiver connection handling failure")
+                .await;
+
+                match res {
+                    Ok(()) => Ok(()),
+                    Err(e) => {
+                        let ok_error = if let Some(pg_error) = e.downcast_ref::<postgres::Error>() {
+                            super::walreceiver_connection::is_expected_error(pg_error)
+                        } else {
+                            e.chain().any(|maybe_pgerr| {
+                                maybe_pgerr
+                                    .downcast_ref::<postgres::Error>()
+                                    .filter(|e| super::walreceiver_connection::is_expected_error(e))
+                                    .is_some()
+                            })
+                        };
```

Comment:
Consider encapsulating this as part of walreceiver_connection.rs under a function such as 
pub fn is_expected_pg_error(error: &anyhow::Error) -> bool

Reply:
I am usually against refactorings for the sake of hiding something ugly.. But I should just chain the `e.downcast_ref` into the any iteration.

Need to re-check if there are anymore callsites for the ignore fn.

Reply:
Moved them into `trait ExpectedError` (scoped to WalReceiver) in [c851071](https://github.com/neondatabase/neon/pull/4402/commits/c85107152448e31c86543f282e2326f1907e10e6) and refactored the if out in [5266a66](https://github.com/neondatabase/neon/pull/4402/commits/5266a66362da80489442c1c1ad9753b691bbe072).